cli/internal/oauth/api/api.go

```diff
@@ -96,17 +96,31 @@ func tryDecodeOAuthError(resp *http.Response) error {
 // authenticated or we have reached the time limit for authenticating (based on
 // the response from GetDeviceCode).
 func (a API) WaitForDeviceToken(ctx context.Context, state State) (TokenResponse, error) {
-	ticker := time.NewTicker(state.IntervalDuration())
+	// Ticker for polling tenant for login – based on the interval
+	// specified by the tenant response.
+	ticker := time.NewTimer(state.IntervalDuration())
 	defer ticker.Stop()
-	timeout := time.After(state.ExpiryDuration())
+	// The tenant tells us for as long as we can poll it for credentials
+	// while the user logs in through their browser. Timeout if we don't get
+	// credentials within this period.
+	timeout := time.NewTimer(state.ExpiryDuration())
+	defer timeout.Stop()
 
 	for {
+		resetTimer(ticker, state.IntervalDuration())
 		select {
 		case <-ctx.Done():
+			// user canceled login
 			return TokenResponse{}, ctx.Err()
 		case <-ticker.C:
+			// tick, check for user login
 			res, err := a.getDeviceToken(ctx, state)
 			if err != nil {
+				if errors.Is(err, context.Canceled) {
```

Comment:
Do we need to handle `context.DeadlineExceeded` here?  We could even use the errdefs helper, but not sure if we want to import it here; https://github.com/docker/cli/blob/092b5f0fcb3e567e616b97ae167c05c42f3b3a1b/vendor/github.com/docker/docker/errdefs/is.go#L120-L123

Reply:
Hmm, the purpose of that:
```go
// if the caller canceled the context, continue
// and let the select hit the ctx.Done() branch
```
is just to make sure we hit the `user canceled login` error if the user cancels execution while `API.getDeviceToken` is going rather than the `context canceled error` from there, so I don't think we need it. This is also all inside `/internal`, so we don't expect others to import this, so I think it's fine. If the user does get a timeout while trying to connect to the tenant, I think we should error out and let them know.

Reply:
Ah, right; yeah, was curious if we wanted to retry if the connection timed out. I guess that would be a bit of an exceptional case, so returning early in that case makes sense as well.

Thanks!